### PR TITLE
[violet] Per-channel surface loss rebalance (adaptive weights)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

The current loss function applies a single global weight to the entire surface prediction (`--surface-loss-weight`) and a single global weight to the volume prediction (`--volume-loss-weight`). Within the surface head, all four channels — `cp` (surface pressure), `wsx`, `wsy`, `wsz` — are weighted equally in the masked MSE.

This is suboptimal because the channels have wildly different learning difficulty:
- `wall_shear_y` (wsy): currently 9.10% vs. 3.65% AB-UPT target — **gap 5.45pp**
- `wall_shear_z` (wsz): currently 10.87% vs. 3.63% AB-UPT target — **gap 7.24pp**
- `surface_pressure` (cp): currently 4.80% vs. 3.82% AB-UPT target — **gap 0.98pp** (close)

The optimizer dedicates equal gradient bandwidth to all four surface channels, but cp converges much faster than wsy/wsz. By the time cp is nearly converged, the optimizer is effectively wasting a large portion of its gradient capacity on a low-residual channel at the expense of the high-gap wsy/wsz axes.

**The fix**: implement per-channel loss reweighting in `train_loss()` using weights inversely proportional to each channel's exponential running mean of the loss value (i.e., uncertainty/difficulty-weighted rebalancing). Hard channels with high residual loss get upweighted; easy channels get downweighted. The weights are updated from running loss statistics and re-normalized at each step so the total surface loss scale stays constant.

This is a continuous, adaptive version of the per-axis reweighting idea, motivated by:
- Uncertainty weighting for multi-task learning (Kendall et al. 2018 — https://arxiv.org/abs/1705.07115)
- The observed wsy/wsz binding constraint on this dataset (gap 5–7pp vs. target)
- Senku's metric-aware-loss result showing vol_p can be decisively won (5.46% < 6.08% target) via explicit loss reformulation

**Prediction**: Per-channel reweighting will reduce the wsy/wsz gap by directing more gradient signal to those axes while not degrading cp or vol_p. If the adaptive weights converge correctly, we expect abupt improvement of 0.5–1.5pp over the alphonse 7.21% baseline with a corresponding improvement specifically in ws_y and ws_z.

## Instructions

You will modify `target/train.py` to add adaptive per-channel surface loss weighting.

### Step 1: Add the `PerChannelLossRebalancer` class

Add the following class near the top of `train.py` (after imports, before `train_loss`):

```python
class PerChannelLossRebalancer:
    """
    Maintains exponential running means of per-channel surface loss values
    and computes normalized per-channel weights inversely proportional to
    the running mean. Channels with higher loss get upweighted; channels
    with lower loss get downweighted. Weights are re-normalized so their
    sum equals the number of channels (preserving the overall loss scale).
    """
    def __init__(self, n_channels: int = 4, momentum: float = 0.99, eps: float = 1e-6):
        self.n_channels = n_channels
        self.momentum = momentum
        self.eps = eps
        # Initialize running means to 1.0 (neutral weighting at step 0)
        self.running_means = torch.ones(n_channels)
        self._initialized = False

    def update(self, per_channel_losses: torch.Tensor) -> torch.Tensor:
        """
        Update running means with current per-channel losses and return
        normalized weights for this step.
        
        Args:
            per_channel_losses: Tensor of shape [n_channels] with current
                                 per-channel loss values (detached).
        Returns:
            weights: Tensor of shape [n_channels], sum = n_channels.
        """
        losses = per_channel_losses.float().cpu()
        if not self._initialized:
            self.running_means = losses.clone()
            self._initialized = True
        else:
            self.running_means = (
                self.momentum * self.running_means + (1.0 - self.momentum) * losses
            )
        # Weights inversely proportional to running mean (harder = higher weight)
        inv = 1.0 / (self.running_means + self.eps)
        weights = inv / inv.mean()  # normalize so mean weight = 1.0 → sum = n_channels
        return weights
```

### Step 2: Add `per_channel_surface_weights` parameter to `train_loss`

Modify the signature of `train_loss` to accept optional per-channel weights:

```python
def train_loss(
    model: nn.Module,
    batch,
    transform: TargetTransform,
    device: torch.device,
    amp_mode: str,
    *,
    surface_loss_weight: float = 1.0,
    volume_loss_weight: float = 1.0,
    raw_rel_l2_weight: float = 0.0,
    per_channel_surface_weights: torch.Tensor | None = None,  # ADD THIS
) -> tuple[torch.Tensor, dict[str, float]]:
```

### Step 3: Apply per-channel weights inside `train_loss`

Replace the surface loss computation block:

```python
surface_loss = masked_mse(out["surface_preds"], surface_target, batch.surface_mask)
```

with:

```python
# Per-channel surface loss
_surface_ch_names = ["cp", "wsx", "wsy", "wsz"]
if per_channel_surface_weights is not None:
    # Compute per-channel MSE separately, then weight
    per_ch_losses = torch.zeros(4, device=device)
    for ci in range(4):
        ch_pred = out["surface_preds"][..., ci:ci+1]
        ch_tgt = surface_target[..., ci:ci+1]
        ch_mask = batch.surface_mask[..., ci:ci+1] if batch.surface_mask.shape[-1] > 1 else batch.surface_mask
        per_ch_losses[ci] = masked_mse(ch_pred, ch_tgt, ch_mask)
    w = per_channel_surface_weights.to(device=device, dtype=per_ch_losses.dtype)
    surface_loss = (per_ch_losses * w).mean()
    # Log per-channel raw losses and weights for diagnostics
    for ci, cname in enumerate(_surface_ch_names):
        aux_metrics[f"per_ch_loss/{cname}"] = float(per_ch_losses[ci].detach().cpu())
        aux_metrics[f"per_ch_weight/{cname}"] = float(w[ci].detach().cpu())
else:
    surface_loss = masked_mse(out["surface_preds"], surface_target, batch.surface_mask)
```

Note: This change requires moving `aux_metrics: dict[str, float] = {}` initialization to **before** the surface_loss computation block (not after). Check the current code layout and adjust accordingly.

### Step 4: Instantiate rebalancer and pass weights in the training loop

In the training loop in `main()`, before the epoch loop, instantiate the rebalancer:

```python
# After: config = parse_args(argv)  # approx line 218
per_ch_rebalancer = PerChannelLossRebalancer(n_channels=4, momentum=0.99)
```

Then in the per-batch training call, update the rebalancer and pass weights:

Find the call to `train_loss(...)` and replace it with:
```python
# Get current channel weights (call before loss, use previous step's weights)
ch_weights = per_ch_rebalancer.running_means  # pass current running means to get weights
# Compute loss with per-channel weighting
loss_val, loss_metrics = train_loss(
    model, batch, transform, device, config.amp_mode,
    surface_loss_weight=config.surface_loss_weight,
    volume_loss_weight=config.volume_loss_weight,
    raw_rel_l2_weight=config.raw_rel_l2_weight,
    per_channel_surface_weights=per_ch_rebalancer.update(
        # Use running means from previous step as this step's weights
        per_ch_rebalancer.running_means.clone()
    ),
)
```

Wait — this needs careful thought. The `update()` method both reads and writes state. Do it cleanly:

```python
# Before train_loss call:
ch_weights = per_ch_rebalancer.update(per_ch_rebalancer.running_means.clone())
# Then pass ch_weights to train_loss; but we also need actual per-channel losses to update.
# Better: use two-step update.
```

Actually, the cleanest implementation: compute weights from **previous step's running means**, then after the backward pass, update the running means with the actual per-channel losses from this step's metrics. Here is the clean pattern:

```python
# Step A: get weights from current (previous-step updated) running means
with torch.no_grad():
    inv = 1.0 / (per_ch_rebalancer.running_means + per_ch_rebalancer.eps)
    ch_weights = inv / inv.mean()

# Step B: call train_loss with these weights
loss_val, loss_metrics = train_loss(
    model, batch, transform, device, config.amp_mode,
    surface_loss_weight=config.surface_loss_weight,
    volume_loss_weight=config.volume_loss_weight,
    raw_rel_l2_weight=config.raw_rel_l2_weight,
    per_channel_surface_weights=ch_weights,
)

# Step C: after computing loss, update running means with this step's actual per-channel losses
# (loss_metrics will contain "per_ch_loss/cp", "per_ch_loss/wsx", etc. when ch_weights is not None)
if "per_ch_loss/cp" in loss_metrics:
    new_losses = torch.tensor([
        loss_metrics["per_ch_loss/cp"],
        loss_metrics["per_ch_loss/wsx"],
        loss_metrics["per_ch_loss/wsy"],
        loss_metrics["per_ch_loss/wsz"],
    ])
    momentum = per_ch_rebalancer.momentum
    if not per_ch_rebalancer._initialized:
        per_ch_rebalancer.running_means = new_losses.clone()
        per_ch_rebalancer._initialized = True
    else:
        per_ch_rebalancer.running_means = (
            momentum * per_ch_rebalancer.running_means + (1.0 - momentum) * new_losses
        )
```

This avoids calling `update()` twice and keeps the logic transparent.

### Step 5: Log per-channel weights to W&B

In the W&B batch logging block, add the per-channel diagnostic metrics from `loss_metrics`:

```python
# After existing W&B logging block, add:
for cname in ["cp", "wsx", "wsy", "wsz"]:
    key_loss = f"per_ch_loss/{cname}"
    key_weight = f"per_ch_weight/{cname}"
    if key_loss in loss_metrics:
        wandb_metrics[f"train/{key_loss}"] = loss_metrics[key_loss]
    if key_weight in loss_metrics:
        wandb_metrics[f"train/{key_weight}"] = loss_metrics[key_weight]
```

### Step 6: DDP consistency

The `per_ch_rebalancer` is a CPU-side running average. In DDP training, each rank sees different batches and will have different running means. This is acceptable — the weights will converge to similar values across ranks given the shared dataset. **Do not all-reduce the running means** (it's unnecessary overhead). The weights are informational; the gradient averaging via DDP handles the actual loss sync.

## Reproduce Command

**Single trial with momentum=0.99** (the proposed default):

```bash
cd target/ && torchrun --standalone --nproc-per-node=4 train.py \
  --model-layers 4 --model-hidden-dim 256 --model-heads 4 \
  --no-use-ema --lr 3e-4 --lr-cosine-t-max 50 --lr-min 1e-6 \
  --no-compile-model \
  --kill-thresholds "500:train/loss<5,12000:val_primary/abupt_axis_mean_rel_l2_pct<22,60000:val_primary/abupt_axis_mean_rel_l2_pct<12,120000:val_primary/abupt_axis_mean_rel_l2_pct<10" \
  --wandb-group bengio-wave3-per-ch-rebalance \
  --wandb-name "violet/per-ch-rebalance-m099" \
  --agent violet
```

Run for 50 epochs. If the per-channel weights diverge (any weight > 10 or < 0.1 in your W&B logs), kill and report immediately.

If the first trial completes and shows improvement in wsy/wsz, run a second trial with `momentum=0.95` (faster adaptation) for comparison:

```bash
cd target/ && torchrun --standalone --nproc-per-node=4 train.py \
  [same flags as above except:]
  --wandb-name "violet/per-ch-rebalance-m095" \
```

(You'll need to expose `momentum` as a CLI flag for this, or hard-code the second trial in the source with a comment.)

## Wave 1 Baseline to Beat

| Metric | Wave 1 best (val) — alphonse PR #74 | AB-UPT target |
|---|---:|---:|
| `val_primary/abupt_axis_mean_rel_l2_pct` | **7.209** | 4.51 |
| `val_primary/surface_pressure_rel_l2_pct` | 4.802 | 3.82 |
| `val_primary/wall_shear_rel_l2_pct` | 8.160 | 7.29 |
| `val_primary/wall_shear_x_rel_l2_pct` | 7.109 | 5.35 |
| `val_primary/wall_shear_y_rel_l2_pct` | 9.100 | 3.65 |
| `val_primary/wall_shear_z_rel_l2_pct` | 10.869 | 3.63 |
| `val_primary/volume_pressure_rel_l2_pct` | **4.166** ✓ | 6.08 |

**Primary metric (lower is better):** `val_primary/abupt_axis_mean_rel_l2_pct`

**Success criteria:** beat 7.209% on abupt AND show wsy < 8.5% and wsz < 9.5% (meaningful gap closure on the binding axes).

## Decision Criteria

- **Merge** if abupt < 7.209% with wsy/wsz both showing gap closure.
- **Request changes** if abupt is 7.2–8.0% but wsy/wsz improved measurably — try momentum=0.95.
- **Close** if abupt > 8.5% at ep20 or per-channel weights diverge (any channel weight > 10x another).

## Diagnostic Checkpoints

Monitor `train/per_ch_weight/{cp,wsx,wsy,wsz}` in W&B from the start. Expected behavior:
- Epoch 1–3: weights start near uniform (1.0 each); wsy/wsz weights begin climbing
- Epoch 5–10: wsy/wsz weights stabilize at ~2–4x cp weight
- If any weight exceeds 8, the rebalancer has over-fit to a stochastic noise spike — report and we'll add gradient clipping or a weight cap

Report per-channel weights at ep1, ep5, ep10, and ep20 alongside the standard val_primary table.

### CODE REVIEW

**Comment by morganmcg1** [comment]
**EP50 COMPLETION — DOES NOT BEAT BASELINE** (2026-04-30 ~14:00 UTC)

Run `em5ixfew` (violet/per-channel-loss-rebalance) has FINISHED at step 890,851 (~ep50).

**Best checkpoint — step 890,849 (~ep50, terminal):**
| Metric | Best val | AB-UPT Target | vs Alphonse |
|--------|:---:|:---:|:---:|
| abupt_axis_mean_rel_l2_pct | **8.449%** | 4.51% | +1.24pp worse |
| surface_pressure_rel_l2_pct | 5.482% | 3.82% | +0.68pp worse |
| volume_pressure_rel_l2_pct | **4.948%** | 6.08% | **BEATS TARGET** ✓ |
| wall_shear_x_rel_l2_pct | 8.168% | 5.35% | +1.06pp worse |
| wall_shear_y_rel_l2_pct | 11.217% | 3.65% | +2.12pp worse |
| wall_shear_z_rel_l2_pct | 12.429% | 3.63% | +1.56pp worse |

**Gate decision: DOES NOT BEAT ALPHONSE BASELINE (7.209%). Closing.**

Analysis: Per-channel loss rebalancing (adaptive weights) at 1.24pp worse than baseline and significantly worse on wsy/wsz — the binding constraint. The adaptive weighting appears to have over-emphasized vol_p at the expense of wall-shear channels. vol_p=4.948% still beats AB-UPT target but this is now confirmed in 4 runs (not a unique contribution). The approach did not help the binding constraint.

This PR is being closed as a dead end. The wsy/wsz binding constraint requires architectural solutions (GradNorm, tangent frame, kNN attention) rather than loss reweighting alone.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #205 because EP50 ASSESSMENT: Run em5ixfew finished at step 890,849 (~ep50). Best val_abupt=8.449% — does not beat alphonse baseline of 7.209%. vol_p=4.948% beats AB-UPT target. However abupt is 1.24pp above baseline. The cosine T_max=50 schedule shows no advantage over T_max=30 for this architecture — consistent with the pattern we see across all Wave 2 runs peaking near ep31. Closing as dead end. Thank you violet..
